File: java/ch01/p1_4.java
/*
    Page 26, problem 1.4.
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class p1_4 {
    private static int bisect_left(List<Integer> seen, int val) {
        int len = seen.size();
        int pos = 0;

        while (pos < len) {
            if (seen.get(pos) >= val)
                break;
            pos++;
        }

        return pos;
    }

    private static int[] smaller_counts(int[] list) {
        int[] counts = new int[list.length];
        List<Integer> seen = new ArrayList<>();
        int last = list.length - 1;

        for (int i = last; i >= 0; i--) {
            int pos = bisect_left(seen, list[i]);
            counts[i] = pos;
            seen.add(pos, list[i]);
        }

        return counts;
    }

    public static void main(String[] args) {
        int[] list;

        if (args.length > 0) {
            list = new int[args.length];
            for (int i = 0; i < args.length; i++) {
                list[i] = Integer.parseInt(args[i]);
            }
        } else {
            list = new int[] {3, 4, 9, 6, 1};
        }

        System.out.println(Arrays.toString(smaller_counts(list)));
    }
}

File: perl/ch01/p1_4.pl
#!/usr/bin/perl

# Page 26, problem 1.4.

use strict;
use warnings;

sub bisect_left {
    my ($val, @list) = @_;
    my $pos = 0;
    my $len = scalar @list;

    while ($pos < $len) {
        last if ($list[$pos] >= $val);
        $pos++;
    }

    return $pos;
}

sub smaller_counts {
    my @list = @_;

    my @seen   = ();
    my @counts = ();
    my @list2  = reverse @list;

    while (defined(my $val = shift @list2)) {
        my $pos = bisect_left($val, @seen);
        push @counts, $pos;
        splice @seen, $pos, 0, $val;
    }

    return reverse @counts;
}

my @result = @ARGV ? smaller_counts(@ARGV) : smaller_counts(3, 4, 9, 6, 1);
local $" = ', ';
print "[@result]\n";

exit;
